### PR TITLE
Decode paths, usernames and passwords.

### PR DESCRIPTION
**Pull request by swankjesse**
A building block towards encoding hostnames.

### CODE REVIEW

File: okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java [hunk 1]
```diff
@@ -285,4 +286,55 @@
...
+        HttpUrl.parse("http://host/%00").decodePathSegments());
+    assertEquals(Arrays.asList("a", "\u2603", "c"),
+        HttpUrl.parse("http://host/a/%E2%98%83/c").decodePathSegments());
+    assertEquals(Arrays.asList("a", "\uD83C\uDF69", "c"),
+        HttpUrl.parse("http://host/a/%F0%9F%8D%A9/c").decodePathSegments());
+    assertEquals(Arrays.asList("a", "b", "c"),
+        HttpUrl.parse("http://host/a/%62/c").decodePathSegments());
+    assertEquals(Arrays.asList("a", "z", "c"),
+        HttpUrl.parse("http://host/a/%7A/c").decodePathSegments());
+    assertEquals(Arrays.asList("a", "z", "c"),
```
Comment (JakeWharton) [inline okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java]: Dupe of above

File: okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java [hunk 2]
```diff
@@ -285,4 +286,55 @@
...
+        HttpUrl.parse("http://host/a/%7a/c").decodePathSegments());
+  }
+
+  @Test public void malformedPercentEncoding() {
```
Comment (JakeWharton) [inline okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java]: Wow, nice.

File: okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java (hunk 1 above)
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java]: (Upper & lowercase)

Comment (JakeWharton) [inline okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java]: Oh, snap. Nice.

**Comment by swankjesse** [comment]
https://github.com/square/okhttp/issues/1486

**Pull request merged**